ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterChangeHandler.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2019 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.clustermap;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.apache.helix.api.listeners.IdealStateChangeListener;
+import org.apache.helix.api.listeners.InstanceConfigChangeListener;
+import org.apache.helix.api.listeners.LiveInstanceChangeListener;
+import org.apache.helix.api.listeners.RoutingTableChangeListener;
+import org.apache.helix.spectator.RoutingTableSnapshot;
+
+
+/**
+ * General handler that handles any resource or state changes in cluster. It exposes API(s) for cluster manager to
+ * access up-to-date cluster infos. Each data center has its own {@link ClusterChangeHandler}.
+ */
+interface ClusterChangeHandler
+    extends InstanceConfigChangeListener, LiveInstanceChangeListener, IdealStateChangeListener,
+            RoutingTableChangeListener {
+
+  /**
+   * Set the initial snapshot in this {@link ClusterChangeHandler}.
+   * @param routingTableSnapshot the snapshot to set
+   */
+  void setRoutingTableSnapshot(RoutingTableSnapshot routingTableSnapshot);
```

Comment:
I agree this is a little bit ugly. We temporarily need this method because the initial routing table change may occur before we add the listener. That is, `ClusterChangeHandler` might miss the initial event and adding listener doesn't trigger a new event. So, in worst case, if `ClusterChangeHandler` may need to wait for `routerUpdater` (within RoutingTableProvider) to periodically refresh routing table snapshot. With this method, we explicitly fetch snapshot from RoutingTableProvider(if initial event already occurs) and set routing table snapshot in `ClusterChangeHandler` which can skip unnecessary wait during initialization.